### PR TITLE
Component integration test support

### PR DESCRIPTION
**Pull request by jeradg**
(Work in progress.)

Adds support for using ember-cli-page-object with component integration tests. Allows for passing a test's `this` context to the page object, which causes the page object to use context-specific versions of actions, queries, etc.

This feature is on the roadmap for ember-cli-page-object v1.1.0.

For discussion see issue https://github.com/san650/ember-cli-page-object/issues/97.

## Todo list

- [x] Support passing test's `this` context to `create()`
- [x] Use test context if available in `find()` and `findWithAssert()` helpers
- [x] Use test context if available in action helpers (`clickable()`, etc.)
- [x] Set page context with `page.setContext(this)`
- [x] Remove page context with `page.removeContext()`
- [x] Add a `render()` helper to call `context.render(hbs...)`
- ~~Make sure page contexts, `setContext()` and `removeContext()` work with pages created with `page.extend({})`~~ (*`extend()` functionality no longer on the immediate roadmap. See https://github.com/san650/ember-cli-page-object/pull/109.*)
- [x] Avoid the need to call `page.removeContext()` after each test, if possible.
- [x] Address Codacy style complaints
- [x] Update `findElementWithAssert()` when a context is present so that it asserts that the element is present (mirroring the acceptance test functionality)
- [x] Add JSDoc for `setContext`, `removeContext` and `render`
- [ ] Add move `render` to `actions` directory (from `context` file`)
- [ ] Add component integration test documentation and examples to GitHub Pages branch
- [ ] Make sure it works with ember-mocha

### CODE REVIEW

File: test-support/page-object/helpers.js [hunk 1]
```diff
@@ -1,27 +1,60 @@
...
+
+export function buildSelector(target, selector, options) {
+  return Ember.$.trim(`${calculateScope(target, options.scope)} ${selector}`);
+}
+
+export function findElementWithAssert(tree, selector, options) {
+  return query(tree, selector, options, findWithAssert);
+}
+
+export function findElement(tree, selector, options, target) {
```
Comment (san650) [inline test-support/page-object/helpers.js]: Codacy Issue found: ['target' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059192/issues/source?bid=2903707&fileBranchId=3076256#l56)

File: test-support/page-object/ceibo.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    return target;
+  }
+};
+
+const DEFAULT_BUILDERS = {
+  descriptor: buildDescriptor,
+  object: buildObject,
+  default: buildDefault
+};
```
Comment (san650) [inline test-support/page-object/ceibo.js]: Codacy Issue found: [Parsing error: Illegal export declaration](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059188/issues/source?bid=2903707&fileBranchId=3076256#l116)

File: test-support/page-object/properties/attribute.js [hunk 3]
```diff
@@ -1,45 +1,31 @@
...
+  return {
+    isDescriptor: true,
+
+    get() {
+      var element = findElementWithAssert(this, selector, options);
 
-  return new Descriptor(getAttribute, options);
+      return element.attr(attributeName);
+    }
+  }
```
Comment (san650) [inline test-support/page-object/properties/attribute.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059193/issues/source?bid=2903707&fileBranchId=3076256#l30)

File: test-support/page-object/properties/clickable.js [hunk 4]
```diff
@@ -37,8 +18,14 @@ function doClick(target, key, options) {
...
+  return {
+    isDescriptor: true,
+
+    get() {
+      click(buildSelector(this, selector, options));
 
-  return new Descriptor(doClick, options);
+      return this;
+    }
+  }
```
Comment (san650) [inline test-support/page-object/properties/clickable.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059195/issues/source?bid=2903707&fileBranchId=3076256#l30)

File: tests/unit/actions/clickable-test.js [hunk 5]
```diff
@@ -1,62 +1,70 @@
...
+    assert.equal(actualSelector, 'span');
   };
 
-  property = buildProperty(clickable(expectedSelector));
+  page = create({
+    foo: clickable('span')
+  });
 
-  property.invoke();
+  page.foo;
```
Comment (san650) [inline tests/unit/actions/clickable-test.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059231/issues/source?bid=2903707&fileBranchId=3076256#l20)

File: tests/unit/actions/clickable-test.js [hunk 6]
```diff
@@ -1,62 +1,70 @@
 import { test } from 'qunit';
 import { moduleFor, buildProperty } from '../test-helper';
```
Comment (san650) [inline tests/unit/actions/clickable-test.js]: Codacy Issue found: ['buildProperty' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059231/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/helpers.js (hunk 1 above)
Comment (san650) [inline test-support/page-object/helpers.js]: Codacy Issue found: ['target' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059192/issues/source?bid=2903707&fileBranchId=3076256#l56)

File: test-support/page-object/ceibo.js (hunk 2 above)
Comment (san650) [inline test-support/page-object/ceibo.js]: Codacy Issue found: [Parsing error: Illegal export declaration](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059188/issues/source?bid=2903707&fileBranchId=3076256#l116)

File: test-support/page-object/properties/attribute.js (hunk 3 above)
Comment (san650) [inline test-support/page-object/properties/attribute.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059193/issues/source?bid=2903707&fileBranchId=3076256#l30)

File: test-support/page-object/properties/clickable.js (hunk 4 above)
Comment (san650) [inline test-support/page-object/properties/clickable.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059195/issues/source?bid=2903707&fileBranchId=3076256#l30)

File: tests/unit/actions/clickable-test.js (hunk 5 above)
Comment (san650) [inline tests/unit/actions/clickable-test.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059231/issues/source?bid=2903707&fileBranchId=3076256#l20)

File: tests/unit/actions/clickable-test.js (hunk 6 above)
Comment (san650) [inline tests/unit/actions/clickable-test.js]: Codacy Issue found: ['buildProperty' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1817059231/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/properties/attribute.js (hunk 3 above)
Comment (san650) [inline test-support/page-object/properties/attribute.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963754/issues/source?bid=2903707&fileBranchId=3076256#l30)

File: tests/unit/actions/clickable-test.js [hunk 7]
```diff
@@ -1,62 +1,104 @@
...
+      page;
 
   window.click = function(actualSelector) {
     assert.equal(actualSelector, expectedSelector);
   };
 
-  property = buildProperty(clickable(expectedSelector));
+  page = create({
+    foo: clickable(expectedSelector)
+  });
```
Comment (san650) [inline tests/unit/actions/clickable-test.js]: Codacy Issue found: [Expected an assignment or function call and instead saw an expression.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963791/issues/source?bid=2903707&fileBranchId=3076256#l20)

File: tests/unit/actions/clickable-test.js [hunk 8]
```diff
@@ -1,62 +1,104 @@
 import { test } from 'qunit';
-import { moduleFor, buildProperty } from '../test-helper';
-import clickable from '../../page-object/properties/clickable';
+import { moduleFor } from '../test-helper';
```
Comment (san650) [inline tests/unit/actions/clickable-test.js]: Codacy Issue found: ['buildProperty' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963791/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/helpers.js [hunk 9]
```diff
@@ -1,26 +1,120 @@
...
+
+    scopes.reverse();
+    scopes.push(targetScope);
+
+    return $.trim(scopes.join(' '));
+  }
+
+  getScopes(node) {
+    var scopes = [];
+
```
Comment (san650) [inline test-support/page-object/helpers.js]: Codacy Issue found: ['target' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963753/issues/source?bid=2903707&fileBranchId=3076256#l56)

File: test-support/page-object/properties/clickable.js [hunk 10]
```diff
@@ -35,10 +14,19 @@ function doClick(target, key, options) {
...
+  return {
+    isDescriptor: true,
+
+    value() {
+      /* global click */
+      click(buildSelector(this, selector, options));
 
-  return new Descriptor(doClick, options);
+      return this;
+    }
```
Comment (san650) [inline test-support/page-object/properties/clickable.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963756/issues/source?bid=2903707&fileBranchId=3076256#l30)

File: test-support/page-object/properties/value.js [hunk 11]
```diff
@@ -1,42 +1,31 @@
-import Descriptor from '../descriptor';
 import { findElementWithAssert, trim } from '../helpers';
```
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ['Descriptor' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963766/issues/source?bid=2903707&fileBranchId=3076256#l1)

File: test-support/page-object/properties/value.js [hunk 12]
```diff
@@ -1,42 +1,31 @@
...
```
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ['trim' is defined but never used.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1879963766/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/properties/collection.js [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+import { create } from '../create';
+import { count } from './count';
+
+function merge(target, ...objects) {
+  objects.forEach(o => Ember.merge(target, o));
+
+  return target;
+}
+
+function select(target, ...keys) {
```
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["select" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1926179177/issues/source?bid=2903707&fileBranchId=3076256#l13)

File: test-support/page-object/properties/collection.js [hunk 14]
```diff
@@ -0,0 +1,149 @@
+import Ember from 'ember';
+import Ceibo from 'ceibo';
```
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1926179177/issues/source?bid=2903707&fileBranchId=3076256#l2)

**Comment by jeradg** [comment]
@san650: I rebased this branch from `one-point-zero`. I think it's possible that we can get these changes in for 1.0.

The build seems to be failing only because the scenario for ember 1.12 fails with the new integration test syntax for rendering templates.

So I'm wondering: Should I rewrite the tests to check that they work with the old component test syntax in Ember 1.12? Or should we just support component tests in >= Ember 1.13?

**Comment by san650** [comment]
@jeradg great work!

I would like for the addon to support Ember 1.12 but I don't think we have to support old component test syntax, so having a check for Ember 1.13+ is a good solution I think.

I still would like to release this after 1.0 so we have less things to test before the release. Is this ok with you?

We were thinking on doing a 1.0 beta release next week (if we have everything ready) to start testing in real projects.

**Comment by jeradg** [comment]
@san650: Waiting until 1.1 is totally OK by me. One thing at a time!

File: test-support/page-object/properties/value.js (hunk 11 above)
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ["trim" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1938584381/issues/source?bid=2903707&fileBranchId=3076256#l1)

File: test-support/page-object/properties/collection.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+import { create } from '../create';
+import { count } from './count';
+
+function merge(target, ...objects) {
+  objects.forEach(o => Ember.merge(target, o));
+
+  return target;
+}
+
+function select(target, ...keys) {
```
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["select" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1938584372/issues/source?bid=2903707&fileBranchId=3076256#l13)

File: test-support/page-object/properties/collection.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
+import Ember from 'ember';
+import Ceibo from 'ceibo';
```
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1938584372/issues/source?bid=2903707&fileBranchId=3076256#l2)

**Comment by jeradg** [comment]
@san650: Tests now pass in Ember 1.12.

Didn't require any changes to the addon. I just needed to check the Ember version in the tests and use alternate template compilation syntax in 1.12.

File: test-support/page-object/properties/value.js (hunk 11 above)
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ["trim" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1947929713/issues/source?bid=2903707&fileBranchId=3076256#l1)

File: test-support/page-object/properties/collection.js (hunk 15 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["select" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1947929704/issues/source?bid=2903707&fileBranchId=3076256#l13)

File: test-support/page-object/properties/collection.js (hunk 16 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1947929704/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/properties/value.js (hunk 11 above)
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ["trim" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1949432792/issues/source?bid=2903707&fileBranchId=3076256#l1)

File: test-support/page-object/properties/collection.js (hunk 15 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["select" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1949432783/issues/source?bid=2903707&fileBranchId=3076256#l13)

File: test-support/page-object/properties/collection.js (hunk 16 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1949432783/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/properties/value.js (hunk 11 above)
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ["trim" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1951043981/issues/source?bid=2903707&fileBranchId=3076256#l1)

File: test-support/page-object/properties/collection.js (hunk 15 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["select" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1951043972/issues/source?bid=2903707&fileBranchId=3076256#l13)

File: test-support/page-object/properties/collection.js (hunk 16 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1951043972/issues/source?bid=2903707&fileBranchId=3076256#l2)

**Comment by jeradg** [comment]
@san650: I tried `setContext(this)` in the `beforeEach()` hook and was running into problems, so I'm not sure if that approach is going to be possible.

**Comment by jeradg** [comment]
@san650: I tried `setContext(this)` in the `beforeEach()` hook and was running into problems, so I'm not sure if that approach is going to be possible. But I think I may have figured out a better way to set the context that fits with the method chaining idiom used in acceptance tests.

You can now do something like:

```js
page.setContext(this)
  .render(hbs`{{my-component}}`)
  .doSomething();
```

If the page object is used in multiple tests, you can still call `page.removeContext()` in the afterEach hook, which works fine.

File: test-support/page-object/properties/collection.js (hunk 15 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["select" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1957769446/issues/source?bid=2903707&fileBranchId=3076256#l13)

File: test-support/page-object/properties/collection.js (hunk 16 above)
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1957769446/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/properties/value.js (hunk 11 above)
Comment (san650) [inline test-support/page-object/properties/value.js]: Codacy Issue found: ["trim" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1957769458/issues/source?bid=2903707&fileBranchId=3076256#l1)

File: test-support/page-object/context.js [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+  removeContext,
+  render
+};
+
+export function bindContextMethods(pageObject) {
+  for (let methodName in methods) {
+    pageObject[methodName] = function() {
+      return methods[methodName].apply(pageObject, arguments);
+    }
+  }
```
Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [Don't make functions within a loop.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1957769436/issues/source?bid=2903707&fileBranchId=3076256#l83)

File: test-support/page-object/context.js [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+const methods = {
+  setContext,
+  removeContext,
+  render
+};
+
+export function bindContextMethods(pageObject) {
+  for (let methodName in methods) {
```
Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1957769436/issues/source?bid=2903707&fileBranchId=3076256#l79)

**Comment by san650** [comment]
@jeradg Looks like github cache is not detecting that most commits are on master already, Can you push the last commit again? maybe a `git commit --amend --no-edit && git push -f origin component-integration-test-ceibo`.

**Comment by jeradg** [comment]
@san650: Done.

File: test-support/page-object/properties/collection.js [hunk 19]
```diff
@@ -1,4 +1,5 @@
 import Ember from 'ember';
+import Ceibo from 'ceibo';
```
Comment (san650) [inline test-support/page-object/properties/collection.js]: Codacy Issue found: ["Ceibo" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/1960669499/issues/source?bid=2903707&fileBranchId=3076256#l2)

File: test-support/page-object/context.js (hunk 17 above)
Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [Don't make functions within a loop.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1960669493/issues/source?bid=2903707&fileBranchId=3076256#l83)

File: test-support/page-object/context.js (hunk 18 above)
Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/san650/ember-cli-page-object/file/1960669493/issues/source?bid=2903707&fileBranchId=3076256#l79)

File: test-support/page-object/context.js (hunk 17 above)
Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [Don't make functions within a loop.](https://www.codacy.com/app/san650/ember-cli-page-object/file/2019082828/issues/source?bid=2903707&fileBranchId=3076256#l83)

Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [Don't make functions within a loop.](https://www.codacy.com/app/san650/ember-cli-page-object/file/2019082828/issues/source?bid=2903707&fileBranchId=3076256#l83)

File: test-support/page-object/context.js (hunk 18 above)
Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/san650/ember-cli-page-object/file/2019082828/issues/source?bid=2903707&fileBranchId=3076256#l79)

Comment (san650) [inline test-support/page-object/context.js]: Codacy Issue found: [The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.](https://www.codacy.com/app/san650/ember-cli-page-object/file/2019082828/issues/source?bid=2903707&fileBranchId=3076256#l79)

File: test-support/page-object/properties/click-on-text.js [hunk 20]
```diff
@@ -45,11 +51,22 @@ export function clickOnText(selector, options = {}) {
...
-        var actualSelector = findChildElement(this, selector, textToClick, options) || findElement(this, selector, textToClick, options);
+      var context = getContext(this);
+
+      if (context) {
+        Ember.run(() => {
+          var fullSelector = actualSelector(this, selector, textToClick, options);
+
+          context.$(fullSelector).click();
+        });
+      } else {
```
Comment (san650) [inline test-support/page-object/properties/click-on-text.js]: Codacy Issue found: ["`wait`" is defined but never used](https://www.codacy.com/app/san650/ember-cli-page-object/file/2020892486/issues/source?bid=2903707&fileBranchId=3076256#l63)

Comment (jeradg) [inline test-support/page-object/properties/click-on-text.js]: This issue seems to be invalid. The `wait` is hit if there's no context.

Comment (san650) [inline test-support/page-object/properties/click-on-text.js]: Yes, sorry about that. Codacy generates these comments automatically based on some predefined rules and it seems it has some issues.

I'll turn off automatic comments from Codacy, it's not that useful.

Comment (san650) [inline test-support/page-object/properties/click-on-text.js]: Looking at it more closely I think the issue is that you wrote wait with back ticks. Try removing the back ticks.

Comment (jeradg) [inline test-support/page-object/properties/click-on-text.js]: Done.

**Comment by jeradg** [comment]
@san650 @juanazam: Component integration test support is ready to go.

I want to write some automated tests to check it with Mocha, but that's probably not worth holding up 1.1 for. (I would also like to add some Mocha tests for the addon as a whole, but there's a bit of tooling that I need to do to allow us to run both QUnit and Mocha tests with Testem.)

**Comment by jeradg** [comment]
Re-pushed to kick the build...

**Comment by san650** [comment]
:tada: :clap:

**Pull request merged**